lib/vkloader.c

```diff
@@ -146,38 +169,179 @@ ktxVulkanDeviceInfo_Construct(ktxVulkanDeviceInfo* This,
+ *                        memory. If NULL, the default allocator will be used.
+ * @param pFunctions      pointer to the struct of functions used for
+ *
+ * @returns KTX_SUCCESS on success, KTX_OUT_OF_MEMORY if a command buffer could
+ *          not be allocated.
+ *
+ * @sa ktxVulkanDeviceInfo_destruct()
+ */
+KTX_error_code
+ktxVulkanDeviceInfo_ConstructEx(ktxVulkanDeviceInfo* This,
+                              VkInstance instance,
+                              VkPhysicalDevice physicalDevice, VkDevice device,
+                              VkQueue queue, VkCommandPool cmdPool,
+                              const VkAllocationCallbacks* pAllocator,
+                              const ktxVulkanFunctions* pFunctions)
+{
 
+#define LOAD_EXT_FUNC(member, fun) \
+do {                             \
+  if ((member).fun == NULL) {            \
+    (member).fun = (PFN_##fun)ktxLoadVulkanFunction(#fun); \
+    if ((member).fun == NULL) {          \
+        return KTX_FALSE;                               \
+    }\
+  } \
+} while (0)
+
+#define LOAD_INST_FUNC(member, instance, fun) \
+do {                             \
+  if ((member).fun == NULL) {            \
+    (member).fun = (PFN_##fun)((member).vkGetInstanceProcAddr)((instance), #fun); \
+    if ((member).fun == NULL) {          \
+        return KTX_FALSE;                               \
+    }\
+  } \
+} while (0)
+
+#define LOAD_DEVICE_FUNC(member, device, fun) \
+do {                             \
+  if ((member).fun == NULL) {            \
+    (member).fun = (PFN_##fun)((member).vkGetDeviceProcAddr)((device), #fun); \
+    if ((member).fun == NULL) {          \
+        return KTX_FALSE;                               \
+    }\
+  } \
+} while (0)
+
+
+    This->instance = instance;
     This->physicalDevice = physicalDevice;
     This->device = device;
     This->queue = queue;
     This->cmdPool = cmdPool;
     This->pAllocator = pAllocator;
-    vkGetPhysicalDeviceMemoryProperties(physicalDevice,
+
+    ktxVulkanFunctions funcs;
+    memset(&funcs, 0, sizeof(ktxVulkanFunctions));
+
+    if (pFunctions != NULL)
+        funcs = *pFunctions;
+
+    if (instance == VK_NULL_HANDLE || pFunctions == NULL) {
+        // This is the old behavior, where no functions where specified. We take a shot at loading them dynamically.
+#if defined(KTX_USE_FUNCPTRS_FOR_VULKAN)
+        // Delay loading not supported so must do it ourselves.
+        if (!ktxVulkanModuleHandle) {
+            ktx_error_code_e kresult = ktxLoadVulkanLibrary();
+            if (kresult != KTX_SUCCESS)
+                return kresult;
+        }
+#endif
+
+        // if pFunctions are null, then we try to load the minimum number of required functions.
+        if (pFunctions == NULL) {
+            LOAD_EXT_FUNC(funcs, vkGetInstanceProcAddr);
+        }
+
+        // If we have no instance, we will need to bind the instance instructions and the function instructions
```

Comment:
I don't understand this comment. Please clarify.

Reply:
Yeah, the English got bad here, will fix 👍